packages/chain-mon/src/faultproof-wd-mon/service.ts

```diff
@@ -0,0 +1,519 @@
+    })
+
+    // Connect to L2.
+    await waitForProvider(this.options.l2RpcProvider, {
+      logger: this.logger,
+      name: 'L2',
+    })
+
+    // Need L2 chain ID to resolve contract addresses.
+    const l2ChainId = await getChainId(this.options.l2RpcProvider)
+    this.l2ChainId = l2ChainId
+
+    // Create the OptimismPortal contract instance. If the optimismPortal option is not provided
+    // then the SDK will attempt to resolve the address automatically based on the L2 chain ID. If
+    // the SDK isn't aware of the L2 chain ID then it will throw an error that makes it clear the
+    // user needs to provide this value explicitly.
+    this.state.portal = getOEContract('OptimismPortal2', l2ChainId, {
+      signerOrProvider: this.options.l1RpcProvider,
+      address: this.options.optimismPortalAddress,
+    })
+
+    // Create the L2ToL1MessagePasser contract instance. If the l2ToL1MessagePasser option is not
+    // provided then we'll use the default address which typically should be correct. It's very
+    // unlikely that any user would change this address so this should work in 99% of cases. If we
+    // really wanted to be extra safe we could do some sanity checks to make sure the contract has
+    // the interface we need but doesn't seem important for now.
+    this.state.messenger = getOEContract('L2ToL1MessagePasser', l2ChainId, {
+      signerOrProvider: this.options.l2RpcProvider,
+      address: this.options.l2ToL1MessagePasserAddress,
+    })
+
+    // Previous versions of wd-mon would try to pick the starting block number automatically but
+    // this had the possibility of missing certain withdrawals if the service was restarted at the
+    // wrong time. Given the added complexity of finding a starting point automatically after FPAC,
+    // it's much easier to simply start a fixed block number than trying to do something fancy. Use
+    // the default configured in this service or use zero if no default is defined.
+    this.state.highestUncheckedBlockNumber = this.options.startBlockNumber
+    if (this.options.startBlockNumber === -1) {
+      this.state.highestUncheckedBlockNumber =
+        DEFAULT_STARTING_BLOCK_NUMBERS[l2ChainId] || 0
+    }
+
+    // Default state is that forgeries have not been detected.
+    this.state.forgeryDetected = false
+    this.state.invalidProposalWithdrawals = []
+    this.state.invalidProofWithdrawals = []
+  }
+
+  // K8s healthcheck
+  async routes(router: ExpressRouter): Promise<void> {
+    router.get('/healthz', async (req, res) => {
+      return res.status(200).json({
+        ok: !this.state.forgeryDetected,
+      })
+    })
+  }
+
+  async main(): Promise<void> {
+    this.metrics.isDetectingForgeries.set(Number(this.state.forgeryDetected))
+    this.metrics.invalidProposalWithdrawals.set(
+      this.state.invalidProposalWithdrawals.length
+    )
+    this.metrics.invalidProofWithdrawals.set(
+      this.state.invalidProofWithdrawals.length
+    )
+
+    for (
```

Comment:
Any particular reason to have this loop backwards?

Reply:
This is because I am removing elements from the array during the loop on the array itself.